File: Amit-Kumar-Singh/Binary_Search/Magnetic_Force_Between_Two_Balls.java
//Time Complexity: O(nlogn + nlogm)
//Space Complexity: O(1)
class Solution {
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position);
        int start = 1;
        int end = (position[position.length-1]-position[0])/(m-1);
        while(start<=end){
            int mid = (start+end)/2;
            int ballCount=1;
            int previous = position[0];
            for(int i = 1;i<position.length;i++){
                if(position[i]-previous>=mid){
                    ballCount++;
                    previous = position[i];
                }
            }
            if(ballCount>=m){
                start = mid+1;
            }
            else{
                end = mid-1;
            }
        }
        return end;
    }
}

File: NikhilKumarGupta/Binary search/Magnetic Force Between Two Balls .cpp
class Solution {
    public:
        int maxDistance(vector<int>& position, int m) {
            sort(position.begin(),position.end());
          int n=position.size();
          int force=0;
          int low=0;
          int high=position.back()-position[0];
          while(low<=high){
            int newposition=position[0];
            int balls=1;
            int mid=(low+high)/2;
    
            for(int i=1;i<n;i++){
                if((position[i]-newposition)>=mid){
                    balls++;
                    newposition=position[i];
                }
    
            }
            if(balls>=m){
                force=mid;
                low=mid+1;
            }
            else{
                high=mid-1;
            }
          }
          return force;
          
        }
    };